homeassistant/components/camera/logi_circle.py

```diff
@@ -0,0 +1,204 @@
+        schema=LOGI_CIRCLE_SERVICE_RECORD)
+
+
+class LogiCam(Camera):
+    """An implementation of a Logi Circle camera."""
+
+    def __init__(self, hass, camera, device_info):
+        """Initialize Logi Circle camera."""
+        super(LogiCam, self).__init__()
+        self._camera = camera
+        self._name = self._camera.name
+
+    @property
+    def name(self):
+        """Return the name of this camera."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Logi Circle camera's support turning on and off ("soft" switch)."""
+        return SUPPORT_ON_OFF
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            'battery_saving_mode': ('on' if self._camera.battery_saving
+                                    else 'off'),
+            'firmware': self._camera.firmware,
+            'generation': self._camera.model_generation,
+            'ip_address': self._camera.ip_address,
+            'mac_address': self._camera.mac_address,
+            'microphone_gain': self._camera.microphone_gain,
+            'model': self._camera.model,
+            'mount': self._camera.mount,
+            'plan': self._camera.plan_name,
+            'speaker_volume': self._camera.speaker_volume,
+            'timezone': self._camera.timezone
+        }
+
+    async def async_camera_image(self):
+        """Return a still image from the camera."""
+        return await self._camera.get_snapshot_image()
+
+    async def async_turn_off(self):
+        """Disable streaming mode for this camera."""
+        await self._camera.set_streaming_mode(False)
+
+    async def async_turn_on(self):
+        """Enable streaming mode for this camera."""
+        await self._camera.set_streaming_mode(True)
+
+    @property
+    def should_poll(self):
+        """Update the image periodically."""
+        return True
+
+    async def set_config(self, mode, value):
+        """Set an configuration property for the target camera."""
+        if mode == LED_MODE_KEY:
+            await self._camera.set_led(value)
+        if mode == PRIVACY_MODE_KEY:
+            await self._camera.set_privacy_mode(value)
+        if mode == BATTERY_SAVING_MODE_KEY:
+            await self._camera.set_battery_saving_mode(value)
+
+    async def download_livestream(self, filename, duration):
+        """Download a recording from the camera's livestream."""
+        # Render filename from template.
+        filename.hass = self.hass
+        stream_file = filename.async_render(
+            variables={ATTR_ENTITY_ID: self.entity_id})
+
+        # Respect configured path whitelist.
+        if not self.hass.config.is_allowed_path(stream_file):
+            _LOGGER.error(
+                "Can't write %s, no access to path!", stream_file)
+            return
+
+        asyncio.shield(self._camera.record_livestream(
```

Comment:
The `record_livestream` method takes a [DASH](https://en.wikipedia.org/wiki/Dynamic_Adaptive_Streaming_over_HTTP) initialization file and concatenates it with stream segments. This uses NamedTemporaryFiles to stage some files, I wanted to guard against cancellation so that those temp files are always cleaned up.

My understanding of asyncio is pretty shaky and I'm not sure what circumstances cancellation would occur. If this pointless let me know and I'll get rid of it. 😄